gsuite_reports_rules/gsuite_drive_external_share.py

```
@@ -0,0 +1,130 @@
+                    current_time < details.get("allowed_until")):
+                return False
+            # No exceptions match.
+            # Return the event summary (which is True) to alert & use in title.
+            return {
+                "actor": actor_email,
+                "doc_title": doc_title,
+                "target_user": target_user,
+            }
+    return False
+
+
+def rule(event):
+    application_name = deep_get(event, 'id', 'applicationName')
+    events = event.get("events")
+    actor_email = deep_get(event, 'actor', 'email', default="EMAIL_UNKNOWN")
+
+    if (application_name == "drive" and events and
+            "acl_change" in set(e["type"] for e in events)):
+        # If any of the events in this record are a dangerous file share, alert:
+        return any(
+            _check_acl_change_event(actor_email, acl_change_event)
+            for acl_change_event in events)
+    return False
+
+
+def title(event):
+    events = event.get("events", [])
+    actor_email = deep_get(event, 'actor', 'email', default='EMAIL_UNKNOWN')
+    matching_events = [
+        _check_acl_change_event(actor_email, acl_change_event)
+        for acl_change_event in events
+        if _check_acl_change_event(actor_email, acl_change_event)
+    ]
+
+    if matching_events:
+        len_events = len(matching_events)
+        first_event = matching_events[0]
+        actor = first_event.get("actor", "ACTOR_UNKNOWN")
+        doc_title = first_event.get("doc_title", "DOC_TITLE_UNKNOWN")
+        target_user = first_event.get("target_user", "USER_UNKNOWN")
+        if len(matching_events) > 1:
+            return f'Multiple dangerous shares ({len_events}) by [{actor}], including "{doc_title}" to {target_user}'
+        return f'Dangerous file share by [{actor}]: "{doc_title}" to {target_user}'
+    # How could there not be if title() gets called?
+    return "ERROR - No matching events, but DangerousShares still fired?"
```

You could also raise an exception here and this would be treated as an erroring rule.

---

Do we have a generic PantherRuleError exception?

---

We define a few exceptions related to using the KV table in `global_helpers/panther_oss_helpers.py`, but we don't have a generic one. Maybe we could add

```class PantherDetectionLogic(Exception):
    """Error returned when a Panther detection encounters an unexpected situation."""
```

Or something like that to either `panther_base_helpers` or `panther_oss_helpers`.